### ISSUE TITLE
Support threading in direct messages?

### ISSUE THREAD

**Issue by kr428** [friendica/friendica#13685]
Currently, integration of Friendica direct messages with Mastodon seems to work usably well. Minor (maybe cosmetic) issue I stumble across then and now: Given how DMs are handled in Mastodon, length restriction included, the messages may end up being threads - in example whenever a Mastodon user decides to at some point respond to an earlier message you sent in a communication flow. All of them seem delivered now so nothing's lost here, but as the direct message view in Friendica isn't hierarchical and has no idea of threads, sometimes messages appear strangely "out of place" and one doesn't see which actual earlier message they relate to. And, slightly worse: It's impossible to actually respond to a particular message directly - I only can respond to that communication with that particular user and have no idea where/how in the Mastodon structure this message will be appear/sorted in. Are there sane ways to improve this...?

**Comment by annando** [comment]
The only sane way is to completely rewrite that part. The best way to describe the code there and the data structure is "legacy".

**Comment by MrPetovan** [comment]
I've long been of the opinion that personally privately addressed (not follower-only for example) threads should be shown in the Private Message page instead of the network. Not sure how it would jive with Diaspora though.

**Comment by annando** [comment]
Isn't it how the system currently is working?

**Comment by MrPetovan** [comment]
I don't know what exactly triggers the creation of a `mail` record. So far I've only been able to receive top-level private messages from Friendica contacts, I've only been able to start private conversations with Mastodon contacts. I can't know if Mastodon contacts have been trying to send me private messages though.

**Comment by annando** [comment]
For Friendica we are using AP, but we transmit a special field to tell that it is a direct message. The same field is used by Pleroma, but not by Mastodon.

**Comment by MrPetovan** [comment]
So it would never be possible to receive a private message in the dedicated section from a Mastodon contact, it would always end up in the Network page?

That's a good reason to have a regular threaded view for private conversations and do away with the Friendica-specific private message screens/behaviors.

**Comment by annando** [comment]
No. We will see private messages from Mastodon as direct message in a case when this is an answer to a direct message from a Friendica user. Also there is a new detection for direct messages, so that most likely now a huge number of private posts will be displayed as direct message.

**Comment by MrPetovan** [comment]
Got it, thanks. The current private message interface is substandard so I would argue against directing more posts its way. A regular network view with a filter on the left to sort on top-level post author/recipient would be an improvement.

**Comment by annando** [comment]
No. We will see private messages from Mastodon as direct message in a case when this is an answer to a direct message from a Friendica user. Also there is a new detection for direct messages, so that most likely now a huge number of private posts will be displayed as direct message.

**Comment by annando** [comment]
We have to rework that whole part, but since Diaspora DMs are using that interface as well, work is needed anyway.

**Comment by HankG** [comment]
The current DM API is a bit wonky too :(.

**Comment by loma-one** [comment]
More functions on the left-hand side? The side navigation is already completely overloaded. 

Is it not possible to create a separate collection point instead? Why not use the navigation point that is currently used by the DM? 
This would fit in well with the chats that are resolved by Firefish or Misskey and processed by Friendica. Private messages have always been a kind of chat.

**Comment by MrPetovan** [comment]
My idea would be a separate network view with different filters in the left column.

**Comment by HankG** [comment]
> My idea would be a separate network view with different filters in the left column.

In terms of how it is implemented in the UI I'm ambivalent. However I absolutely hate how DMs and regular post streams are all merged together on Mastodon. It seems like a horrible UX.

**Comment by MrPetovan** [comment]
The ideal for me would be a chat-like display where all the messages to and from the same recipients are flattened into a single stream like in instant messages apps. However, I feel like it would be more work than reusing the existing Network paradigm, so the latter is my compromise position.
